File: 0798-smallest-rotation-with-highest-score/0798-smallest-rotation-with-highest-score.py
class Solution:
    def bestRotation(self, nums):
        n = len(nums)
        change = [0] * (n + 1)

        for i, num in enumerate(nums):
            left = (i - num + 1 + n) % n
            right = (i + 1) % n

            change[left] -= 1
            change[right] += 1

            if left > right:
                change[0] += 1

        best = 0
        score = 0
        maxScore = -1

        for i in range(n):
            score += change[i]
            if score > maxScore:
                maxScore = score
                best = i

        return best

File: 0798-smallest-rotation-with-highest-score/0798-smallest-rotation-with-highest-score.java
class Solution {
    public int bestRotation(int[] nums) {
        int n = nums.length;
        int[] change = new int[n + 1];

        for (int i = 0; i < n; i++) {
            int num = nums[i];
            int left = (i - num + 1 + n) % n;
            int right = (i + 1) % n;

            change[left]--;
            change[right]++;

            if (left > right) {
                change[0]++;
            }
        }

        int best = 0;
        int score = 0;
        int maxScore = -1;

        for (int i = 0; i < n; i++) {
            score += change[i];
            if (score > maxScore) {
                maxScore = score;
                best = i;
            }
        }

        return best;
    }
}